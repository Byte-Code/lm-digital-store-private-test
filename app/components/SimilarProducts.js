import React, { Component } from 'react';
import ImmutablePropTypes from 'react-immutable-proptypes';
import { List } from 'immutable';
import styled from 'styled-components';

import ProductBadge from './ProductBadge';
import SimilarProductsDialog from './SimilarProductsDialog';

const Wrapper = styled.div`
  padding-left: 40px;
  display: flex;
`;

const Header = styled.div`
  width: 90px;
  &>h3 {
    font-size: 16px;
    text-transform: uppercase;
    padding-right: 60px;
  }
`;

const Slider = styled.div`
  display: flex;
  overflow-x: auto;
  &>div {
    margin-right: 20px;
  }
`;

export default class SimilarProducts extends Component {
  static propTypes = {
    similarProducts: ImmutablePropTypes.list,
  }

  static defaultProps = {
    similarProducts: List()
  }

  constructor(props) {
    super(props);
    this.state = {
      dialogOpen: false,
      selectedProduct: ''
    };
  }

  handleOpen = (product) => {
    this.setState({
      selectedProduct: product,
      dialogOpen: true
    });
  }

  handleClose = () => {
    this.setState({ dialogOpen: false });
  }

  renderProducts() {
    const { similarProducts } = this.props;

    return similarProducts.map(p => (
      <ProductBadge
        key={p.get('code')}
        productInfo={p}
        handleClick={() => this.handleOpen(p.get('code'))}
      />
    ));
  }

  render() {
    const { similarProducts } = this.props;
    if (similarProducts.isEmpty()) {
      return null;
    }
    const { dialogOpen, selectedProduct } = this.state;

    return (
      <Wrapper>
        <Header>
          <h3>prodotti simili</h3>
        </Header>
        <Slider>
          {this.renderProducts()}
        </Slider>
        <SimilarProductsDialog
          similarProducts={similarProducts}
          handleClose={this.handleClose}
          isOpen={dialogOpen}
          selectedProduct={selectedProduct}
        />
      </Wrapper>
    );
  }
}
