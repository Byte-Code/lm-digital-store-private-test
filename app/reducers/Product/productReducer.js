import { Map } from 'immutable';

export default function productReducer(state = Map(), action) {
  const { type, productCode, result } = action;
  switch (type) {
    case 'SUCCESS_FETCH_PRODUCT':
      return state.set(productCode, result);
    case 'SUCCESS_FETCH_SIMILARPRODUCTS':
      return state.setIn([productCode, 'similarProducts'], result);
    default:
      return state;
  }
}
