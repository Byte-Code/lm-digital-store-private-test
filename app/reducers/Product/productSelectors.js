export default function getProduct(state, productCode) {
  return state.get(productCode);
}
