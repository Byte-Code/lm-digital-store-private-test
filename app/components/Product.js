import React, { Component, PropTypes } from 'react';
import ImmutablePropTypes from 'react-immutable-proptypes';
import { Map } from 'immutable';
import styled from 'styled-components';

import ImageSlider from './ImageSlider';
import ProductInfo from './ProductInfo';
import PriceBadge from './PriceBadge';
import SimilarProducts from './SimilarProducts';

const Wrapper = styled.div`
  position: relative;
`;

const Title = styled.h1`
  margin: 40px 40px 0;
  text-align: center;
  font-size: 48px;
  line-height: 70px;
  text-transform: capitalize;
`;

const Ref = styled.h3`
  text-transform: uppercase;
  font-size: 16px;
  line-height: 24px;
  text-align: center;
  margin-bottom: 16px;
`;

const SliderWrapper = styled.div`
  width: 100%;
`;

const PriceWrapper = styled.div`
  position: absolute;
  right: 30px;
  top: 234px;
`;

const SimilarProductsWrapper = styled.div`
  margin: 60px 0 80px;
`;

export default class Product extends Component {
  static propTypes ={
    params: PropTypes.shape({
      productCode: PropTypes.string.isRequired
    }).isRequired,
    productInfo: ImmutablePropTypes.map,
    requestFetchProduct: PropTypes.func.isRequired
  }

  static defaultProps = {
    productInfo: Map()
  }

  componentDidMount() {
    const { params: { productCode }, requestFetchProduct } = this.props;
    requestFetchProduct(productCode);
  }

  componentDidUpdate(prevProps) {
    const prevProductCode = prevProps.params.productCode;
    const { params: { productCode }, requestFetchProduct } = this.props;

    if (prevProductCode !== productCode) {
      requestFetchProduct(productCode);
    }
  }

  render() {
    const { productInfo } = this.props;

    if (productInfo.isEmpty()) {
      return null;
    }

    const name = productInfo.get('name');
    const code = productInfo.get('code');
    const productType = productInfo.getIn(['productDetail', 'descriptionType']);
    const marketingDescriptions = productInfo.getIn(['productDetail', 'marketingDescriptions']);
    const descriptions = productInfo.getIn(['productDetail', 'descriptions']);
    const similarProducts = productInfo.get('similarProducts');
    const pricingInfo = productInfo.getIn(['price', 'selling']);
    // TODO this data should't arrive from here, selector Maybe?
    const currentStoreStock = productInfo.get('storeStock');
    const allStoreStock = productInfo.get('allStoreStock');
    const imageIDList = productInfo.get('images');
    const imageOptions = { width: 1080, height: 1080, crop: 'fit' };

    return (
      <Wrapper>
        <Title>{name}</Title>
        <Ref>{`REF. ${code}`}</Ref>
        <SliderWrapper>
          <ImageSlider
            imageIDList={imageIDList}
            imageOptions={imageOptions}
            alt={name}
          />
        </SliderWrapper>
        <ProductInfo
          productType={productType}
          marketingDescriptions={marketingDescriptions}
          descriptions={descriptions}
        />
        <SimilarProductsWrapper>
          <SimilarProducts
            similarProducts={similarProducts}
          />
        </SimilarProductsWrapper>
        <PriceWrapper>
          <PriceBadge
            productName={name}
            productCode={code}
            pricingInfo={pricingInfo}
            currentStoreStock={currentStoreStock}
            allStoreStock={allStoreStock}
          />
        </PriceWrapper>
      </Wrapper>
    );
  }
}
