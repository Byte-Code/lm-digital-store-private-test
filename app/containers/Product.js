import { connect } from 'react-redux';

import Product from '../components/Product';
import { requestFetchProduct } from '../actions/productActions';
import { clearProductList } from '../actions/productListActions';
import { getProduct, getStoreCode, getSimilarProducts } from '../reducers/selectors';

const mapStateToProps = (state, ownProps) => {
  const { params: { productCode } } = ownProps;
  return {
    productInfo: getProduct(state, productCode),
    storeCode: getStoreCode(state),
    similarProducts: getSimilarProducts()(state, productCode)
  };
};

const mapDispatchToProps = {
  requestFetchProduct,
  clearProductList
};

export default connect(mapStateToProps, mapDispatchToProps)(Product);
