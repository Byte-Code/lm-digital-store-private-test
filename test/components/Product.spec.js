import React from 'react';
import { shallow } from 'enzyme';
import { fromJS, List, Map } from 'immutable';

import mountWithStore from '../../app/utils/testingUtils';
import Product from '../../app/components/Product';
import SimilarProducts from '../../app/components/SimilarProducts';

const store = fromJS({
  storeReducer: {
    name: 'store'
  }
});

const params = { productCode: 'PRODUCT123' };
const requestFetchProduct = jest.fn();
const storeCode = '89';
const emptyProductInfo = Map();
const emptySimilarProducts = List();
const productInfo = fromJS({
  name: 'name',
  code: 'code',
  slug: 'slug',
  productDetail: {
    descriptionType: 'descriptionType',
    marketingDescriptions: {},
    descriptions: []
  },
  marketingAttributes: {},
  price: { selling: {} },
  pricingInformations: {},
  allStoreStock: [],
  images: ['image1'],
  loyaltyProgram: {},
  similarProducts: [
    {
      name: 'anche',
      products: ['prod1', 'prod2']
    }
  ]
});
const similarProducts = fromJS([{ code: 'prod1' }, { code: 'prod3' }]);

describe('Product', () => {
  it('should return null when productInfo is empty', () => {
    const result = shallow(
      <Product
        productInfo={emptyProductInfo}
        params={params}
        hasNearbyStores
        storeCode={storeCode}
        similarProducts={emptySimilarProducts}
        requestFetchProduct={requestFetchProduct}
        clearProductList={() => {}}
      />
    );
    expect(result).toMatchSnapshot();
  });

  it('should render properly with no similarProducts ', () => {
    const result = shallow(
      <Product
        productInfo={productInfo}
        params={params}
        hasNearbyStores
        storeCode={storeCode}
        similarProducts={emptySimilarProducts}
        requestFetchProduct={requestFetchProduct}
        clearProductList={() => {}}
      />
    );
    expect(result.find(SimilarProducts)).toHaveLength(0);
    expect(result).toMatchSnapshot();
  });

  it('should render properly with similarProducts ', () => {
    const result = shallow(
      <Product
        productInfo={productInfo}
        params={params}
        hasNearbyStores
        storeCode={storeCode}
        similarProducts={similarProducts}
        requestFetchProduct={requestFetchProduct}
        clearProductList={() => {}}
      />
    );
    expect(result.find(SimilarProducts)).toHaveLength(1);
    expect(result).toMatchSnapshot();
  });

  it('should call requestFetchProduct after mounting', () => {
    mountWithStore(
      <Product
        productInfo={productInfo}
        params={params}
        hasNearbyStores
        storeCode={storeCode}
        similarProducts={emptySimilarProducts}
        requestFetchProduct={requestFetchProduct}
        clearProductList={() => {}}
      />,
      store
    );
    expect(requestFetchProduct).toHaveBeenCalledWith(params.productCode);
  });

  it('should call requestFetchProduct again when productCodeChanges', () => {
    const result = mountWithStore(
      <Product
        productInfo={productInfo}
        params={params}
        hasNearbyStores
        storeCode={storeCode}
        similarProducts={emptySimilarProducts}
        requestFetchProduct={requestFetchProduct}
        clearProductList={() => {}}
      />,
      store
    );
    result.setProps({ params: { productCode: 'PRODUCT456' } });
    expect(requestFetchProduct).toHaveBeenCalledWith('PRODUCT456');
  });
});
