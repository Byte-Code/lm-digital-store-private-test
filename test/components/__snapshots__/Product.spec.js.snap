// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product should render properly with no similarProducts  1`] = `
<glamorous(div)>
  <glamorous(h1)>
    name
  </glamorous(h1)>
  <glamorous(h3)>
    REF. code
  </glamorous(h3)>
  <glamorous(div)>
    <ImageSlider
      alt="name"
      imageIDList={
        Array [
          "image1",
        ]
      }
      imageOptions={
        Object {
          "crop": "fit",
          "height": 1080,
          "width": 1080,
        }
      }
    />
  </glamorous(div)>
  <ProductInfo
    descriptions={Array []}
    marketingDescriptions={Object {}}
    productType="descriptionType"
  />
  <glamorous(div) />
  <glamorous(div)>
    <ProductInfoBadge
      currentStoreStock={
        Object {
          "stockStatus": undefined,
          "storeStock": undefined,
        }
      }
      hasNearbyStores={true}
      loyaltyProgram={Object {}}
      marketingAttributes={Object {}}
      price={Object {}}
      pricingInfo={Object {}}
      productCode="code"
      productName="name"
      productSlug="slug"
    />
  </glamorous(div)>
</glamorous(div)>
`;

exports[`Product should render properly with similarProducts  1`] = `
<glamorous(div)>
  <glamorous(h1)>
    name
  </glamorous(h1)>
  <glamorous(h3)>
    REF. code
  </glamorous(h3)>
  <glamorous(div)>
    <ImageSlider
      alt="name"
      imageIDList={
        Array [
          "image1",
        ]
      }
      imageOptions={
        Object {
          "crop": "fit",
          "height": 1080,
          "width": 1080,
        }
      }
    />
  </glamorous(div)>
  <ProductInfo
    descriptions={Array []}
    marketingDescriptions={Object {}}
    productType="descriptionType"
  />
  <glamorous(div)>
    <SimilarProducts
      similarProducts={
        Array [
          Object {
            "code": "prod1",
          },
        ]
      }
      title="anche"
    />
  </glamorous(div)>
  <glamorous(div)>
    <ProductInfoBadge
      currentStoreStock={
        Object {
          "stockStatus": undefined,
          "storeStock": undefined,
        }
      }
      hasNearbyStores={true}
      loyaltyProgram={Object {}}
      marketingAttributes={Object {}}
      price={Object {}}
      pricingInfo={Object {}}
      productCode="code"
      productName="name"
      productSlug="slug"
    />
  </glamorous(div)>
</glamorous(div)>
`;

exports[`Product should return null when productInfo is empty 1`] = `null`;
